File: gramatica_portuguesa.pl

/*
sentenca --> noun_phrase, verbal_phrase.
sentenca --> noun_phrase_p, verbal_phrase_p.

frase_nom --> determiner_f, noun_f.
frase_nom --> determiner_m, noun_m.
frase_nom --> noun_f.
frase_nom --> noun_m.

frase_nom_p --> determiner_p_f, noun_p_f.
frase_nom_p --> determiner_p_m, noun_p_m.
frase_nom_p --> noun_p_f.
frase_nom_p --> noun_p_m.
*/

%Nome e verbo
sentenca(sentenca(N,V)) --> frase_nom(N), frase_ver(V).
sentenca(sentenca(NP,VP)) --> frase_nom_p(NP), frase_ver_p(VP).

%Nome  artigo ou nome -- Tudo no singular
frase_nom(frase_nom(Nome)) --> nome_m(Nome).
frase_nom(frase_nom(Nome)) --> nome_f(Nome).

frase_nom(frase_nom(Artigo, Nome)) --> artigo_m(Artigo), nome_m(Nome).
frase_nom(frase_nom(Artigo, Nome)) --> artigo_f(Artigo), nome_f(Nome).

frase_nom(frase_nom(Contra, Nome)) --> contracao_m(Contra), nome_m(Nome).
frase_nom(frase_nom(Contra, Nome)) --> contracao_f(Contra), nome_f(Nome).


%Nome artigo ou nome -- Tudo Plural
frase_nom_p(frase_nom_p(Nome)) --> nome_p_m(Nome).
frase_nom_p(frase_nom_p(Nome)) --> nome_p_f(Nome).

frase_nom_p(frase_nom_p(Artigo, Nome)) --> artigo_p_m(Artigo), nome_p_m(Nome).
frase_nom_p(frase_nom_p(Artigo, Nome)) --> artigo_p_f(Artigo), nome_p_f(Nome).

frase_nom_p(frase_nom_p(Contra, Nome)) --> contracao_p_m(Contra), nome_p_m(Nome).
frase_nom_p(frase_nom_p(Contra, Nome)) --> contracao_p_f(Contra), nome_p_f(Nome).

%Preposição
frase_prep(frase_prep(Preposi,Frase)) --> prep(Preposi), frase_nom(Frase).
frase_prep(frase_prep(Preposi,Frase)) --> prep(Preposi), frase_nom_p(Frase).

%Verbo singular
frase_ver(frase_ver(Verbo,Resto)) --> verbo(Verbo), frase_nom(Resto).
frase_ver(frase_ver(Verbo,Resto)) --> verbo(Verbo), frase_nom_p(Resto).
frase_ver_p(frase_ver_p(Verbo,Resto)) --> verbo(Verbo), frase_prep(Resto).
frase_ver(frase_ver(Verbo)) --> verbo(Verbo).

%Verbo plural
frase_ver_p(frase_ver_p(Verbo,Resto)) --> verbo_p(Verbo), frase_nom(Resto).
frase_ver_p(frase_ver_p(Verbo,Resto)) --> verbo_p(Verbo), frase_nom_p(Resto).
frase_ver_p(frase_ver_p(Verbo,Resto)) --> verbo_p(Verbo), frase_prep(Resto).
frase_ver_p(frase_ver_p(Verbo)) --> verbo_p(Verbo).

%Agora anar a fazer levantamentos de palavras

%nome masculino
nome_m(nome(cacador)) --> [cacador].
nome_m(nome(cachorro)) --> [cachorro].
nome_m(nome(rio)) --> [rio].
nome_m(nome(mar)) --> [mar].
nome_m(nome(tempo)) --> [tempo].
nome_m(nome(sino)) --> [sino].
nome_m(nome(vento)) --> [vento].
nome_m(nome(martelo)) --> [martelo].
nome_m(nome(tambor)) --> [tambor].
nome_m(nome(tempo)) --> [tempo].
nome_m(nome(rosto)) --> [rosto].

%nome femenino
nome_f(nome(menina)) --> [menina].
nome_f(nome(noticia)) --> [noticia].
nome_f(nome(floresta)) --> [floresta].
nome_f(nome(porta)) --> [porta].
nome_f(nome(vida)) --> [vida].
nome_f(nome(mae)) --> [mae].
nome_f(nome(cidade)) --> [cidade].

%plurais
nome_p_m(nome(lobos)) --> [lobos].
nome_p_m(nome(tambores)) --> [tambores].

nome_p_f(nome(meninas)) --> [meninas].
nome_p_f(nome(lagrimas)) --> [lagrimas].

%verbos
verbo(verbo(corre)) --> [corre].
verbo(verbo(correu)) --> [correu].
verbo(verbo(bateu)) --> [bateu].

verbo_p(verbo_p(corriam)) --> [corriam].
verbo_p(verbo_p(bateram)) --> [bateram].
verbo_p(verbo_p(correram)) --> [correram].

%preposicoes
prep(preposicao(para)) --> [para].
prep(preposicao(com)) --> [com].

%contracoes femininas
contracao_f(artigo(na)) --> [na].
contracao_f(artigo(pela)) --> [pela].


%contracoes masculinas
contracao_m(artigo(no)) --> [no].
contracao_m(artigo(pelo)) --> [pelo].

%plural
contracao_p_m(artigo(nos)) --> [nos].
contracao_p_m(artigo(pelos)) --> [pelos].

contracao_p_f(artigo(nas)) --> [nas].
contracao_p_f(artigo(pelas)) --> [pelas].

%artigo masculo
artigo_m(artigo(o)) --> [o].
artigo_m(artigo('O')) --> ['O'].

%artugo femenino
artigo_f(artigo(a)) --> [a].
artigo_f(artigo('A')) --> ['A'].

%plurais
artigo_p_m(artigo(os)) --> [os].
artigo_p_m(artigo('Os')) --> ['Os'].

artigo_p_f(artigo(as)) --> [as].
artigo_p_f(artigo('As')) --> ['As'].

File: gramatica_portuguesa.py
from sys import argv

nomeM = ['cacador', 'cachorro', 'rio', 'mar', 'tempo', 'sino', 'vento', 'martelo', 'tambor', 'tempo','rosto']
nomeF = ['menina','noticia', 'floresta','porta','vida','mae','cidade']

nomePM = ['lobos','tambores']
nomePF = ['meninas','lagrimas']

verboS = ['corre','correu','bateu']
verboP = ['corriam', 'bateram','correram']
preposicao = ['para', 'com']

contracaoM = ['no','pelo']
contracaoF = ['na','pela']

contracaoPM = ['nos','pelos']
contracaoPF = ['nas','pelas']

artigoF = ['a','A']
artigoM = ['o','O']

artigoPM = ['os','Os']
artigoPF = ['as','As']

carta = {'nomeM':['artigoM','contracaoM'],'nomeF':['artigoF','contracaoF'],'nomePM':['artigoPM','contracaoPM'],'nomePF':['artigoPF','contracaoPF']}

def preparaString(frase_raw):
    frase_preparada = frase_raw
    tipos = []
    for i in frase_preparada:
        if i in preposicao:
            tipos.append('preposicao')
        elif i in contracaoM:
            tipos.append('contracaoM')
        elif i in contracaoF:
            tipos.append('contracaoF')
        elif i in contracaoPM:
            tipos.append('contracaoPM')
        elif i in contracaoPF:
            tipos.append('contracaoPF')
        elif i in artigoM:
            tipos.append('artigoM')
        elif i in artigoF:
            tipos.append('artigoF')
        elif i in artigoPM:
            tipos.append('artigoPM')
        elif i in artigoPF:
            tipos.append('artigoPF')
        elif i in verboS:
            tipos.append('verboS')
        elif i in verboP:
            tipos.append('verboP')
        elif i in nomeM:
            tipos.append('nomeM')
        elif i in nomeF:
            tipos.append('nomeF')
        elif i in nomePM:
            tipos.append('nomePM')
        elif i in nomePF:
            tipos.append('nomePF')
        else:
            raise ValueError('Nome não identificado!')
    return frase_preparada, tipos


def fraseNomial(componentes,tipo, generic=0):
    if generic == 0:
        targets = ['nomeM','nomeF','nomePM','nomePF']
        for i in range(4):
            num, ret = r_fraseNomial(componentes,tipo,targets[i])
            if num != 0 and i < 2:
                return num, ret, 's'
            elif num != 0:
                return num, ret, 'p'
        return 0, None, None
    else:
        if generic == 1:
            for i in ['nomeM','nomeF']:
                num, ret = r_fraseNomial(componentes, tipo, i)
                if num != 0:
                    return num, ret, None
            return 0, None, None
        else:
            for i in ['nomePM','nomePF']:
                num, ret = r_fraseNomial(componentes, tipo, i)
                if num != 0:
                    return num, ret, None
            return 0, None, None


def r_fraseNomial(componentes,tipo,target):
    global carta
    teste = carta[target][0]
    if tipo[0] == target:
        return 1, 'fraseNominal(nome('+componentes[0]+')'')'
    elif len(componentes) >= 2 and tipo[0] == carta[target][0] and tipo[1] == target:
        return 2, 'fraseNominal(artigo('+componentes[0]+')nome('+componentes[1]+'))'
    elif len(componentes) >= 2 and tipo[0] == carta[target][1] and tipo[1] == target:
        return 2, 'fraseNominal(contracao(' + componentes[0] + ')nome(' + componentes[1] + '))'
    else:
        return 0, None


def frasePreposicional(componentes, tipos):
    if tipos[0] == 'preposicao':
        num, ret, _ = fraseNomial(componentes[1:],tipos[1:])
        if num == 0:
            raise ValueError('Frase incorreta!')
        return 1+num, 'frasePreposicional(preposicao('+componentes[0]+')'+ret+')'
    else:
        return 0, None

def fraseVerbal(componentes, tipos, type):
    if type == 's':
        flag = 1 # singular
        target = 'verboS'
    else:
        flag = 2 # plural
        target = 'verboP'

    if tipos[0] == target:
        if len(componentes)>1:
            num, ret, _ = fraseNomial(componentes[1:],tipos[1:],flag)
            if num != 0:
                return 1 + num, 'fraseVerbal(verbo(' + componentes[0] + ')' + ret + ')'
            else:
                num, ret = frasePreposicional(componentes[1:], tipos[1:])
                if num != 0:
                    return 1+num, 'fraseVerbal(verbo(' + componentes[0] + ')'+ret+')'
                else:
                    return 1, 'fraseVerbal(verbo('+componentes[0]+'))'
        else:
            return 1, 'fraseVerbal(verbo(' + componentes[0] + '))'
    else:
        return 0, None


componentes, tipos = preparaString(argv[1:])

'''Sentença'''
num , ret, flag = fraseNomial(componentes, tipos)
if num == 0:
    raise ValueError('Frase incorreta!')
resultado = ret

num, ret = fraseVerbal(componentes[num:],tipos[num:], flag)
if num == 0:
    raise ValueError('Frase incorreta!')
print('sent('+resultado+ret+')')